File: dotnet/DesignPatterns/StructuralDesignPatterns/ProxyDesignPattern/ProxyDesignPattern.DynamicProxyExample/LoggingProxy.cs
using System.Diagnostics;
using System.Reflection;

namespace ProxyDesignPattern.DynamicProxyExample;

/// <summary>
/// Dynamic Proxy using DispatchProxy (built-in .NET).
/// Automatically logs all method calls on any interface at runtime.
/// No need to create a separate proxy class for each interface!
/// </summary>
public class LoggingProxy<T> : DispatchProxy where T : class
{
    private T? _target;

    /// <summary>
    /// Creates a dynamic proxy that wraps the target object.
    /// </summary>
    public static T Create(T target)
    {
        // Create proxy instance
        var proxy = Create<T, LoggingProxy<T>>();
        
        // Set the target
        ((LoggingProxy<T>)(object)proxy)._target = target;
        
        return proxy;
    }

    /// <summary>
    /// Intercepts all method calls on the proxy.
    /// </summary>
    protected override object? Invoke(MethodInfo? targetMethod, object?[]? args)
    {
        if (targetMethod == null || _target == null)
            return null;

        var methodName = targetMethod.Name;
        var argsString = args != null ? string.Join(", ", args) : "";

        // Before method execution
        Console.WriteLine($"[Proxy] Calling: {typeof(T).Name}.{methodName}({argsString})");
        var stopwatch = Stopwatch.StartNew();

        try
        {
            // Invoke the actual method
            var result = targetMethod.Invoke(_target, args);

            stopwatch.Stop();

            // After method execution
            Console.WriteLine($"[Proxy] Returned: {result} (took {stopwatch.ElapsedMilliseconds}ms)");

            return result;
        }
        catch (TargetInvocationException ex)
        {
            stopwatch.Stop();
            Console.WriteLine($"[Proxy] Exception: {ex.InnerException?.Message} (took {stopwatch.ElapsedMilliseconds}ms)");
            throw ex.InnerException ?? ex;
        }
    }
}

File: nodejs/DesignPatterns/StructuralDesignPatterns/ProxyDesignPattern/ProxyDesignPattern.DynamicProxyExample/src/LoggingProxy.ts
// ============================================================================
// DYNAMIC PROXY USING JAVASCRIPT'S NATIVE Proxy OBJECT
// ============================================================================
// JavaScript's built-in Proxy is the direct equivalent of .NET's DispatchProxy.
// It intercepts all property accesses and method calls at runtime.
//
// Benefits:
//   - No need to create a separate proxy class for each interface
//   - One createLoggingProxy<T> works for ANY object
//   - Add cross-cutting concerns (logging, timing, etc.) dynamically
// ============================================================================

/**
 * Creates a dynamic logging proxy that wraps any object.
 * Intercepts all method calls, logs arguments and return values, and measures timing.
 *
 * Uses JavaScript's native Proxy object — the equivalent of .NET's DispatchProxy.
 */
export function createLoggingProxy<T extends object>(target: T, interfaceName: string): T {
  return new Proxy(target, {
    get(obj: T, prop: string | symbol) {
      const value = (obj as Record<string | symbol, unknown>)[prop];

      // Only intercept function calls
      if (typeof value !== "function") {
        return value;
      }

      return function (...args: unknown[]) {
        const methodName = String(prop);
        const argsString = args.join(", ");

        // Before method execution
        console.log(`[Proxy] Calling: ${interfaceName}.${methodName}(${argsString})`);
        const start = Date.now();

        try {
          // Invoke the actual method
          const result = (value as (...a: unknown[]) => unknown).apply(obj, args);

          const elapsed = Date.now() - start;
          console.log(`[Proxy] Returned: ${result} (took ${elapsed}ms)`);

          return result;
        } catch (err: unknown) {
          const elapsed = Date.now() - start;
          const message = err instanceof Error ? err.message : String(err);
          console.log(`[Proxy] Exception: ${message} (took ${elapsed}ms)`);
          throw err;
        }
      };
    },
  });
}
